tls/s2n_ktls.c

```diff
@@ -297,3 +300,146 @@ int s2n_config_ktls_enable_unsafe_tls13(struct s2n_config *config)
+{
+    RESULT_ENSURE_REF(estimate);
+    uint64_t records = bytes / S2N_TLS_MAXIMUM_FRAGMENT_LENGTH;
+    if (bytes % S2N_TLS_MAXIMUM_FRAGMENT_LENGTH) {
+        records++;
+    }
+    *estimate = records;
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_set_estimated_sequence_number(
+        struct s2n_connection *conn, size_t bytes_written)
+{
+    RESULT_ENSURE_REF(conn);
+    if (conn->actual_protocol_version < S2N_TLS13) {
+        return S2N_RESULT_OK;
+    }
+
+    uint64_t new_records_sent = 0;
+    RESULT_GUARD(s2n_ktls_estimate_records(bytes_written, &new_records_sent));
+
+    struct s2n_blob seq_num = { 0 };
+    RESULT_GUARD(s2n_connection_get_sequence_number(conn, conn->mode, &seq_num));
+
+    for (size_t i = 0; i < new_records_sent; i++) {
+        RESULT_GUARD_POSIX(s2n_increment_sequence_number(&seq_num));
+    }
+    return S2N_RESULT_OK;
+}
+
+/* We need to track when the key encryption limit is reached. We could get the current record
+ * sequence number from the kernel with getsockopt, but that requires a surprisingly
+ * expensive syscall.
+ *
+ * Instead, we track the estimated sequence number and enforce the limit based
+ * on that estimate.
+ */
+S2N_RESULT s2n_ktls_check_estimated_record_limit(struct s2n_connection *conn, size_t bytes_requested)
+{
+    RESULT_ENSURE_REF(conn);
+    if (conn->actual_protocol_version < S2N_TLS13) {
+        return S2N_RESULT_OK;
+    }
+
+    uint64_t new_records_sent = 0;
+    RESULT_GUARD(s2n_ktls_estimate_records(bytes_requested, &new_records_sent));
+
+    uint64_t old_records_sent = 0;
+    struct s2n_blob seq_num = { 0 };
+    RESULT_GUARD(s2n_connection_get_sequence_number(conn, conn->mode, &seq_num));
+    RESULT_GUARD_POSIX(s2n_sequence_number_to_uint64(&seq_num, &old_records_sent));
+
+    RESULT_ENSURE(S2N_ADD_IS_OVERFLOW_SAFE(old_records_sent, new_records_sent, UINT64_MAX),
+            S2N_ERR_KTLS_KEY_LIMIT);
+    uint64_t total_records_sent = old_records_sent + new_records_sent;
+
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg);
+    uint64_t encryption_limit = conn->secure->cipher_suite->record_alg->encryption_limit;
+    if (total_records_sent > encryption_limit) {
+        RESULT_ENSURE_REF(conn->config);
+        RESULT_ENSURE(conn->config->ktls_tls13_enabled, S2N_ERR_KTLS_KEY_LIMIT);
+
+        /* Check that the ktls send call is not going over the encryption limit
+         * more than once, as that would require multiple key updates.
+         * We can add this logic in the future, but it's not required at the moment. */
+        RESULT_ENSURE(S2N_ADD_IS_OVERFLOW_SAFE(encryption_limit, encryption_limit, UINT64_MAX),
+                S2N_ERR_KTLS_KEY_LIMIT);
+        uint64_t ktls_encryption_limit = encryption_limit * 2;
+        RESULT_ENSURE(new_records_sent <= ktls_encryption_limit, S2N_ERR_INVALID_ARGUMENT);
```

Comment:
Is this math right? What if old_records_sent = (encryption_limit - 1) and new_records_sent = (encryption_limit + 3)?

Shouldn't it be total_records_sent <= encryption_limit*2?

Reply:
Darn, you're right. I didn't consider you might be under the record limit, and go over the record limit twice with the new records sent, but your total new records is smaller than encryption limit * 2. Lemme fix this...